File: client/src/test/scala/stasis/test/specs/unit/client/ops/backup/stages/MetadataPushSpec.scala
package stasis.test.specs.unit.client.ops.backup.stages

import akka.actor.ActorSystem
import akka.stream.{Materializer, SystemMaterializer}
import akka.stream.scaladsl.{Sink, Source}
import stasis.client.analysis.Checksum
import stasis.client.api.clients.Clients
import stasis.client.encryption.secrets.DeviceSecret
import stasis.client.model.{DatasetMetadata, FilesystemMetadata}
import stasis.client.ops.backup.Providers
import stasis.client.ops.backup.stages.MetadataPush
import stasis.shared.model.datasets.DatasetDefinition
import stasis.shared.ops.Operation
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.client.Fixtures
import stasis.test.specs.unit.client.mocks.{MockBackupTracker, _}

class MetadataPushSpec extends AsyncUnitSpec { spec =>
  "A Backup MetadataPush stage" should "push dataset metadata" in {
    val mockEncryption = new MockEncryption()
    val mockApiClient = MockServerApiEndpointClient()
    val mockCoreClient = MockServerCoreEndpointClient()
    val mockTracker = new MockBackupTracker

    val stage = new MetadataPush {
      override protected def targetDataset: DatasetDefinition = Fixtures.Datasets.Default
      override protected def deviceSecret: DeviceSecret = Fixtures.Secrets.Default
      override protected def providers: Providers =
        Providers(
          checksum = Checksum.MD5,
          staging = new MockFileStaging(),
          compression = MockCompression(),
          encryptor = mockEncryption,
          decryptor = mockEncryption,
          clients = Clients(
            api = mockApiClient,
            core = mockCoreClient
          ),
          track = mockTracker,
          telemetry = MockClientTelemetryContext()
        )

      override implicit protected def mat: Materializer = SystemMaterializer(system).materializer
    }

    implicit val operationId: Operation.Id = Operation.generateId()

    Source
      .single(
        DatasetMetadata(
          contentChanged = Map(
            Fixtures.Metadata.FileOneMetadata.path -> Fixtures.Metadata.FileOneMetadata
          ),
          metadataChanged = Map(
            Fixtures.Metadata.FileTwoMetadata.path -> Fixtures.Metadata.FileTwoMetadata,
            Fixtures.Metadata.FileThreeMetadata.path -> Fixtures.Metadata.FileThreeMetadata
          ),
          filesystem = FilesystemMetadata(
            entities = Map(
              Fixtures.Metadata.FileOneMetadata.path -> FilesystemMetadata.EntityState.New,
              Fixtures.Metadata.FileTwoMetadata.path -> FilesystemMetadata.EntityState.Updated,
              Fixtures.Metadata.FileThreeMetadata.path -> FilesystemMetadata.EntityState.Updated
            )
          )
        )
      )
      .via(stage.metadataPush)
      .runWith(Sink.ignore)
      .map { _ =>
        mockEncryption.statistics(MockEncryption.Statistic.FileEncrypted) should be(0)
        mockEncryption.statistics(MockEncryption.Statistic.FileDecrypted) should be(0)
        mockEncryption.statistics(MockEncryption.Statistic.MetadataEncrypted) should be(1)
        mockEncryption.statistics(MockEncryption.Statistic.MetadataDecrypted) should be(0)

        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetEntryCreated) should be(1)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetEntryRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetEntryRetrievedLatest) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetEntriesRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetDefinitionCreated) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetDefinitionRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetDefinitionsRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.PublicSchedulesRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.PublicScheduleRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetMetadataWithEntryIdRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DatasetMetadataWithEntryRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.UserRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.DeviceRetrieved) should be(0)
        mockApiClient.statistics(MockServerApiEndpointClient.Statistic.Ping) should be(0)

        mockCoreClient.statistics(MockServerCoreEndpointClient.Statistic.CratePulled) should be(0)
        mockCoreClient.statistics(MockServerCoreEndpointClient.Statistic.CratePushed) should be(1)

        mockTracker.statistics(MockBackupTracker.Statistic.EntityDiscovered) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.SpecificationProcessed) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.EntityExamined) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.EntityCollected) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.EntityProcessingStarted) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.EntityPartProcessed) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.EntityProcessed) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.MetadataCollected) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.MetadataPushed) should be(1)
        mockTracker.statistics(MockBackupTracker.Statistic.FailureEncountered) should be(0)
        mockTracker.statistics(MockBackupTracker.Statistic.Completed) should be(0)
      }
  }

  private implicit val system: ActorSystem = ActorSystem(name = "MetadataPushSpec")
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/ops/backup/stages/MetadataPushSpec.kt
package stasis.test.client_android.lib.ops.backup.stages

import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.collect
import stasis.client_android.lib.analysis.Checksum
import stasis.client_android.lib.api.clients.Clients
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.FilesystemMetadata
import stasis.client_android.lib.model.server.datasets.DatasetDefinition
import stasis.client_android.lib.ops.Operation
import stasis.client_android.lib.ops.backup.Providers
import stasis.client_android.lib.ops.backup.stages.MetadataPush
import stasis.test.client_android.lib.Fixtures
import stasis.test.client_android.lib.mocks.MockBackupTracker
import stasis.test.client_android.lib.mocks.MockCompression
import stasis.test.client_android.lib.mocks.MockEncryption
import stasis.test.client_android.lib.mocks.MockFileStaging
import stasis.test.client_android.lib.mocks.MockServerApiEndpointClient
import stasis.test.client_android.lib.mocks.MockServerCoreEndpointClient

class MetadataPushSpec : WordSpec({
    "A Backup MetadataPush stage" should {
        "push dataset metadata" {
            val mockEncryption = MockEncryption()
            val mockApiClient = MockServerApiEndpointClient()
            val mockCoreClient = MockServerCoreEndpointClient()
            val mockTracker = MockBackupTracker()

            val stage = object : MetadataPush {
                override val targetDataset: DatasetDefinition = Fixtures.Datasets.Default

                override val deviceSecret: DeviceSecret = Fixtures.Secrets.Default

                override val providers: Providers = Providers(
                    checksum = Checksum.Companion.MD5,
                    staging = MockFileStaging(),
                    compressor = MockCompression(),
                    encryptor = mockEncryption,
                    decryptor = mockEncryption,
                    clients = Clients(
                        api = mockApiClient,
                        core = mockCoreClient
                    ),
                    track = mockTracker
                )
            }

            val metadata = DatasetMetadata(
                contentChanged = mapOf(
                    Fixtures.Metadata.FileOneMetadata.path to Fixtures.Metadata.FileOneMetadata
                ),
                metadataChanged = mapOf(
                    Fixtures.Metadata.FileTwoMetadata.path to Fixtures.Metadata.FileTwoMetadata,
                    Fixtures.Metadata.FileThreeMetadata.path to Fixtures.Metadata.FileThreeMetadata
                ),
                filesystem = FilesystemMetadata(
                    entities = mapOf(
                        Fixtures.Metadata.FileOneMetadata.path to FilesystemMetadata.EntityState.New,
                        Fixtures.Metadata.FileTwoMetadata.path to FilesystemMetadata.EntityState.Updated,
                        Fixtures.Metadata.FileThreeMetadata.path to FilesystemMetadata.EntityState.Updated
                    )
                )
            )

            stage
                .metadataPush(operation = Operation.generateId(), flow = listOf(metadata).asFlow())
                .collect()

            mockEncryption.statistics[MockEncryption.Statistic.FileEncrypted] shouldBe (0)
            mockEncryption.statistics[MockEncryption.Statistic.FileDecrypted] shouldBe (0)
            mockEncryption.statistics[MockEncryption.Statistic.MetadataEncrypted] shouldBe (1)
            mockEncryption.statistics[MockEncryption.Statistic.MetadataDecrypted] shouldBe (0)

            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetEntryCreated] shouldBe (1)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetEntryRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetEntryRetrievedLatest] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetEntriesRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetDefinitionCreated] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetDefinitionRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetDefinitionsRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.PublicSchedulesRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.PublicScheduleRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetMetadataWithEntryIdRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DatasetMetadataWithEntryRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.UserRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.DeviceRetrieved] shouldBe (0)
            mockApiClient.statistics[MockServerApiEndpointClient.Statistic.Ping] shouldBe (0)

            mockCoreClient.statistics[MockServerCoreEndpointClient.Statistic.CratePulled] shouldBe (0)
            mockCoreClient.statistics[MockServerCoreEndpointClient.Statistic.CratePushed] shouldBe (1)

            mockTracker.statistics[MockBackupTracker.Statistic.EntityExamined] shouldBe (0)
            mockTracker.statistics[MockBackupTracker.Statistic.EntityCollected] shouldBe (0)
            mockTracker.statistics[MockBackupTracker.Statistic.EntityProcessed] shouldBe (0)
            mockTracker.statistics[MockBackupTracker.Statistic.MetadataCollected] shouldBe (0)
            mockTracker.statistics[MockBackupTracker.Statistic.MetadataPushed] shouldBe (1)
            mockTracker.statistics[MockBackupTracker.Statistic.FailureEncountered] shouldBe (0)
            mockTracker.statistics[MockBackupTracker.Statistic.Completed] shouldBe (0)
        }
    }
})
